src/internal/components/portal/index.tsx

```diff
@@ -5,29 +5,57 @@ import { createPortal } from 'react-dom';
 
 export interface PortalProps {
   container?: null | Element;
+  getContainer?: () => Promise<HTMLElement>;
+  removeContainer?: (container: HTMLElement) => void;
   children: React.ReactNode;
 }
 
+function manageDefaultContainer(setState: React.Dispatch<React.SetStateAction<Element | null>>) {
+  const newContainer = document.createElement('div');
+  document.body.appendChild(newContainer);
+  setState(newContainer);
+  return () => {
+    document.body.removeChild(newContainer);
+  };
+}
+
+function manageAsyncContainer(
+  getContainer: () => Promise<HTMLElement>,
+  removeContainer: (container: HTMLElement) => void,
+  setState: React.Dispatch<React.SetStateAction<Element | null>>
+) {
+  let newContainer: HTMLElement;
+  getContainer().then(
+    container => {
+      newContainer = container;
+      setState(container);
+    },
+    error => {
+      console.warn('[AwsUi] [portal]: failed to load portal root', error);
+    }
+  );
+  return () => {
+    removeContainer(newContainer);
+  };
+}
+
 /**
  * A safe react portal component that renders to a provided node.
  * If a node isn't provided, it creates one under document.body.
  */
-export default function Portal({ container, children }: PortalProps) {
+export default function Portal({ container, getContainer, removeContainer, children }: PortalProps) {
   const [activeContainer, setActiveContainer] = useState<Element | null>(container ?? null);
 
   useLayoutEffect(() => {
     if (container) {
       setActiveContainer(container);
       return;
     }
-    const newContainer = document.createElement('div');
-    document.body.appendChild(newContainer);
-    setActiveContainer(newContainer);
-    return () => {
-      document.body.removeChild(newContainer);
-      setActiveContainer(null);
-    };
-  }, [container]);
+    if (getContainer && removeContainer) {
```

Comment:
Do you think it would be helpful to customers to raise a warning in case only one of them is provided?

Reply:
Done